### ISSUE TITLE
\ character inside JSX text flags no-useless-escape

### ISSUE THREAD

**Issue by sstern6** [eslint/eslint#7460]
This bug was confirmed by @kaicataldo in `babel-eslint`.  https://github.com/babel/babel-eslint/issues/414
**Tell us about your environment**
- **ESLint Version:** 3.6.0
- **Node Version:** 6.5.0
- **npm Version:** 3.10.3

**What parser (default, Babel-ESLint, etc.) are you using?** `Babel`

**Please show your full configuration:**

```
{
    "parserOptions": {
        "ecmaFeatures": {
            "jsx": true
        }
    },
    "parser": "babel-eslint",
    "rules": {
      "no-useless-escape": ["error"]
    }
}
```

**What did you do? Please include the actual source code causing the issue.**

``` js
<div>Here it is: \ </div>
```

**What did you expect to happen?**
No errors

**What actually happened? Please include the actual, raw output from ESLint.**

`error  Unnecessary escape character: \  no-useless-escape`

**Comment by mysticatea** [comment]
Thank you for this issue and that PR.

I confirmed it with espree as well.

**Comment by nzakas** [comment]
@sstern6 can you explain why this is a bug? It's not clear to me why this is a bad thing.

**Comment by ljharb** [comment]
@nzakas in the context of react, that text is rendered as literal text - ie, it's not escaping a space, it's a literal backslash followed by a literal space.

**Comment by kellyselden** [comment]
@nzakas When 3.9 came out, I started getting these:

```
/xxx1.js
  102:43  error  Unnecessary escape character: \-  no-useless-escape

/xxx2.js
  77:57  error  Unnecessary escape character: \.  no-useless-escape
  77:59  error  Unnecessary escape character: \?  no-useless-escape

/xxx3.js
  569:61  error  Unnecessary escape character: \.  no-useless-escape
```

The samples are:

```js
let reg = /^[a-zA-Z]+[a-zA-Z._\-]*$/;

if (subject !== 'undefined' && !/[\w\s\-\.\?\\!]$/.test(subject)) {

cost = confirmText.match(/\$([0-9]+[\.]*[0-9]*)/)[0],
```

This is Ember.js code.

**Comment by platinumazure** [comment]
@ljharb Does that apply to *any* use of JSX, or just React? I don't think we can make specific exceptions for React semantics here.

**Comment by ljharb** [comment]
@platinumazure any use - since the React team defines the jsx spec, I'm not aware of any deviations wrt the jsx syntax from what React supports.

**Comment by platinumazure** [comment]
@kellyselden This issue is about JSX, which you're not using in your examples. It is entirely intentional that `no-useless-escape` flags escapes that are useless in character classes.

Specific to your examples:

* `\-` is useless in a character class if it's at one end or the other, it only needs to be escaped in the middle.
* `\.` and `\?` do not need to be escaped in character classes at all. (Applies to your second and third example)

So the rule is working as designed in all cases.

In future, please be careful to create a new issue if your issue isn't the same as the issue you're otherwise posting in. (And ideally, please search for open or closed issues that already cover your problem if you can.)

**Comment by not-an-aardvark** [comment]
@kellyselden That is working as intended (and unrelated to this issue), because those characters are uselessly escaped. To fix the issue, remove the backslashes before those characters, and your code will still work correctly. 3.9 included a bugfix where useless escapes like this weren't getting caught by the rule.

Also see: https://github.com/eslint/eslint/issues/7495, https://github.com/eslint/eslint/issues/7507

**Comment by kellyselden** [comment]
@platinumazure OK, thank you for the heads up. I will delete my comment to get rid of the noise.

**Comment by mysticatea** [comment]
@nzakas This is simply false positive because JSX syntax does not have escape sequences. JSX has HTML entities instead. no-useless-escape is ignoring JSXAttribute. It needs to ignore JSXText as well.

**Comment by nzakas** [comment]
@ljharb thanks for explaining, I wasn't aware of this.

@platinumazure JSX is a separate specification from React, so its behavior is well-defined. React applies semantics on top of JSX.

Just a reminder: please be sure to include relevant information (like that the slash isn't an escape in JSX) in bug reports. Not everyone has such context in their brains. :)
